tools/cleanup_duplicate_host_guest_devices.rb

```diff
@@ -45,6 +46,20 @@
 
 return if opts[:dry_run]
 
-GuestDevice.destroy(guest_devices_to_delete) unless opts[:dry_run]
+dependents = GuestDevice.reflect_on_all_associations(:has_one) + GuestDevice.reflect_on_all_associations(:has_many)
+dependent_destroys = dependents.select { |assoc| assoc.options[:dependent] == :destroy }
+
+total_slices = (guest_devices_to_delete.count / opts[:page_size].to_f).ceil
+guest_devices_to_delete.each_slice(opts[:page_size]).with_index do |slice, index|
+  puts "Destroying slice #{index + 1} of #{total_slices}..."
+  GuestDevice.delete(slice)
+
+  dependent_destroys.each do |dependent_assoc|
+    foreign_key = dependent_assoc.options[:foreign_key] || "guest_device_id"
+    dependent_assoc.klass.where(foreign_key => slice).destroy_all
```

Comment:
note: this is N+1 ish

```ruby
klass.where(:id => slice).to_a.each(&:destroy)
```

could we use a `delete` here?

Reply:
If we called this style of batch destroy recursively then yeah but otherwise we'd be leaving possibly orphaned records on those just kicking the can down one level

Reply:
what a great idea :)

Reply:
Hahah yeah we might as well contribute it back to rails at that point